### ISSUE TITLE
After calling DeleteTopics, ReadPartitions(topic) still returns partitions

### ISSUE THREAD

**Issue by justinrush** [segmentio/kafka-go#627]
**Describe the bug**
If I create the topic, sleep a bit, and then issue the delete command, future calls to ReadPartitions still return partitions for that topic. I see the topics being deleted in the Kafka log:
```
 kafka [2021-04-06 13:46:19,870] INFO [Partition default-2-0 broker=0] Log loaded for partition default-2-0 with initial high watermark 0 (kafka.cluster.Partition)
 kafka [2021-04-06 13:47:19,804] INFO [Log partition=default-2-0, dir=/bitnami/kafka/data] Deleting segments as the log has been deleted: LogSegment(baseOffset=0, size=0, lastModifiedTime=1617716179592, largestRecordTimestamp=None)
 kafka [2021-04-06 13:47:19,805] INFO [Log partition=default-2-0, dir=/bitnami/kafka/data] Deleting segment files LogSegment(baseOffset=0, size=0, lastModifiedTime=1617716179592, largestRecordTimestamp=None) (kafka.log.Log)
kafka [2021-04-06 13:47:19,806] INFO Deleted log /bitnami/kafka/data/default-2-0.0cd75ffd43f84009b5e4a50332398ca1-delete/00000000000000000000.log.deleted. (kafka.log.LogSegment)
kafka [2021-04-06 13:47:19,806] INFO Deleted offset index /bitnami/kafka/data/default-2-0.0cd75ffd43f84009b5e4a50332398ca1-delete/00000000000000000000.index.deleted. (kafka.log.LogSegment)
kafka [2021-04-06 13:47:19,806] INFO Deleted time index /bitnami/kafka/data/default-2-0.0cd75ffd43f84009b5e4a50332398ca1-delete/00000000000000000000.timeindex.deleted. (kafka.log.LogSegment)
kafka [2021-04-06 13:47:19,806] INFO Deleted log for partition default-2-0 in /bitnami/kafka/data/default-2-0.0cd75ffd43f84009b5e4a50332398ca1-delete. (kafka.log.LogManager)  
```

**Kafka Version**
What version(s) of Kafka are you testing against?
2.7.0

**To Reproduce**
Steps to reproduce the behavior.  Bonus points for a code sample.
Deploy kafka using the [bitnami helm chart](https://github.com/bitnami/charts/tree/master/bitnami/kafka) with these values:
```yaml
defaultReplicationFactor: 2
deleteTopicEnable: true
externalAccess:
  autoDiscovery:
    enabled: true
  enabled: true
rbac:
  create: true
replicaCount: 3
```

Use this code to create a topic and "wait" for it to be deleted.

```golang
func main() {
	brokers := []string{"kafka-0:9094", "kafka-1:9094", "kafka-2:9094"}
	topic := "default-2"

	conn, err := kafka.Dial("tcp", brokers[0])
	if err != nil {
		log.Fatalf("unable to connect to kafka: %v", err)
	}
	defer conn.Close()

	controller, err := conn.Controller()
	if err != nil {
		log.Fatalf("unable to get brokers: %v", err)
	}

	var controllerConn *kafka.Conn
	controllerConn, err = kafka.Dial("tcp", net.JoinHostPort(controller.Host, strconv.Itoa(controller.Port)))
	if err != nil {
		log.Fatalf("unable to get controller conn: %v", err)
	}
	defer controllerConn.Close()

	topicConfigs := []kafka.TopicConfig{
		kafka.TopicConfig{
			Topic:             topic,
			NumPartitions:     3,
			ReplicationFactor: 2,
		},
	}

	err = controllerConn.CreateTopics(topicConfigs...)
	if err != nil {
		log.Fatalf("unable to create topic: %v", err)
	}

	log.Infof("created %s", topic)
	time.Sleep(10 * time.Second)

	err = controllerConn.DeleteTopics(topic)
	if err != nil {
		log.Fatalf("error calling delete: %v", err)
	}

	for {
		parts, err := controllerConn.ReadPartitions(topic)
		if err != nil {
			if errors.Is(err, kafka.LeaderNotAvailable) {
				log.Infof("leader not available getting parts, trying again")
				time.Sleep(1 * time.Millisecond)

				continue
			}
			log.Errorf("error getting parts: %v", err)
			break
		}

		log.Infof("%s exists with %d parts", topic, len(parts))
		time.Sleep(1 * time.Second)
	}
}
```
**Expected behavior**
I expect to see none of the:
```
time="2021-04-06T08:46:25-05:00" level=info msg="default-2 exists with 1 parts"
```
log lines, or just a few. However, what happens is that it always prints out. I can kill the program and restart it an hour later (commenting out the create and delete bits) and still see that it finds partitions.

**Additional context**
Ultimately I want a way to "await" the DeleteTopics command so I know when it is actually gone. If this isn't the right way, please point me in the right direction.

**Comment by godwhoa** [comment]
From Java client's docs, might be relavent for kafka-go as well

> Delete a batch of topics. This operation is not transactional so it may succeed for some topics while fail for others. It may take several seconds after the DeleteTopicsResult returns success for all the brokers to become aware that the topics are gone. During this time, AdminClient#listTopics and AdminClient#describeTopics may continue to return information about the deleted topics. If delete.topic.enable is false on the brokers, deleteTopics will mark the topics for deletion, but not actually delete them. The futures will return successfully in this case. This operation is supported by brokers with version 0.10.1.0 or higher.

**Comment by achille-roussel** [comment]
@godwhoa you are correct. This is unrelated to kafka-go, it is feature of Kafka itself; topic creation and deletion is asynchronous.

@justinrush if your program needs stronger guarantees, it has to implement them. For example, polling for DescribeTopics after a DeleteTopics until the topic you have deleted does not appear anymore. You can even push one step further and run that query on all brokers to wait for the state to propagate throughout the entire cluster. I don't know if Kafka provides guarantees on when the propagation will occur, in my experience the state tends to converge within a few seconds.

_TLDR: distributed systems are hard._

**Comment by justinrush** [comment]
I don't see a DescribeTopics - is that just the ReadPartitions method? I'm fine implementing some polling method, I'd just like to know the source-of-truth to poll and some guarantee that it will actually eventually do what I asked it to do. If ReadParitions is the source of truth (good to know against each broker), then I can do that, but the above code never seems to show it deleted

**Comment by achille-roussel** [comment]
Sorry, I corrected my message to replace DescribeTopics for ReadPartitions.

kafka.Conn.ReadPartitions should work, or kafka.Client.Metadata https://pkg.go.dev/github.com/segmentio/kafka-go#Client.Metadata, both use the same Kafka API underneath to fetch the state of the broker.
